typing/ctype.ml

```diff
@@ -1210,20 +1215,21 @@ let instance_list schl =
   For_copy.with_scope (fun copy_scope ->
     List.map (fun t -> copy copy_scope t) schl)
 
-let reified_var_counter = ref Vars.empty
-let reset_reified_var_counter () =
-  reified_var_counter := Vars.empty
-
-(* names given to new type constructors.
-   Used for existential types and
-   local constraints *)
-let get_new_abstract_name s =
-  let index =
-    try Vars.find s !reified_var_counter + 1
-    with Not_found -> 0 in
-  reified_var_counter := Vars.add s index !reified_var_counter;
-  if index = 0 && s <> "" && s.[String.length s - 1] <> '$' then s else
-  Printf.sprintf "%s%d" s index
+(* Create unique names to new type constructors.
+   Used for existential types and local constraints. *)
+let get_new_abstract_name env s =
+  (* unique names are needed only for error messages *)
+  if in_counterexample () then s else
+  let rec loop index =
+    let name =
+      if index = 0 && s <> "" && s.[String.length s - 1] <> '$' then s else
+      Printf.sprintf "%s%d" s index
+    in
+    match Env.find_type_by_name (Longident.Lident name) env with
+    | _ -> loop (index + 1)
+    | exception Not_found -> name
+  in
```

Comment:
I was a bit worried by the quadratic worst case of the logic here when called repeatedly on the same constructor. (I can see a pattern matching on a deeply nested GADT list for example.) I propose #12034 to alleviate this. (We don't need to change the present PR, but we want to eventually send a follow-up change to use `Misc.find_first_mono` in this code.)

Reply:
Done.